### ISSUE TITLE
Ruby: Provide reset mechanism for low-level resources

### ISSUE THREAD

**Issue by blowmage** [grpc/grpc#8798]
Please add a mechanism to be called on the Ruby GRPC client to reset low-level resources. This will be useful when the Ruby GRPC client is in a process that has been forked and resources need to be reset. This will help when Ruby GRPC is used in Rails apps using Puma or Unicorn, which regularly fork processes.

In #7951 @soltanmm-google said:

> Forking processes and using gRPC across processes is not supported behavior due to very low-level resource issues. Short of having a way to introspect the kernel resources we're using and micromanaging them from some wrapped call to fork in a system-dependent way (which... just... no, heck no) this isn't tractable.

The Ruby GRPC client does not need to micromanage kernel resources to detect when its process has been forked, as this can be called by the user on process fork. All we need is a mechanism to do so.

**Comment by murgatroid99** [comment]
@ctiller @nicolasnoble What would we need to do to create this?

**Comment by soltanmm-google** [comment]
If Ruby provides a way to just *drop* the module s.t. it's entirely unloaded, and if Ruby's wrapper calls `grpc_shutdown` deterministically on module clean-up, then that should be sufficient, and gRPC doesn't need to do anything (or really even should), right?

**Comment by jrun** [comment]
Is there anything I can do to help move this issue forward? This issue prevents me from being able to use grpc dependent google-cloud-ruby libraries in my internal services.

**Comment by apolcyn** [comment]
@jrun this is under discussion but there are some difficulties around it. 

But getting some more specific use cases and problems would be helpful. For example, do you need "pre-fork" and "port-fork" hooks to reset an active gRPC library? Possibly deferred library startup is sufficient?

**Comment by jrun** [comment]
Thanks for following up. In our specific use case a  "post-fork" callback is needed to reset active gRPC connections. Unfortunately a deferred library startup isn’t sufficient.

Our backend Ruby services use a prefork model. A post-fork callback is used to re-establish network connections (e.g. database, messaging, caching services) in the forked process. A deferred library startup isn’t sufficient because the master process often needs to access the same APIs as the child processes. 

For example the master may initially read it’s configuration from Cloud Datastore to determine the set of child workers to spawn. The child workers then may need to read/write to Cloud Datastore as part of their operation. 

Another example is Stackdriver Error Reporting. The master and child workers need to be able to write to the Error Reporting service.

**Comment by jrun** [comment]
I want to clarify one point in an effort to avoid any miscommunication. We don't need the gRPC library to be provided a "post-fork" callback. We need a method to call that re-establishes the underlying gRPC connections which will be called from our own "post-fork" callback.

**Comment by jrun** [comment]
@apolcyn Is there anything additional I can provide? Does the use case I provide make sense?

**Comment by apolcyn** [comment]
@jrun thanks for data point, this is helpful and makes sense. AFAICS supporting the case described here will require some complicated changes to the core C-library that grpc-ruby is wrapping. But this is important - taking a look at how feasible this is.

**Comment by Gubbi** [comment]
@apolcyn Is there a temporary workaround recommended until this is fixed at the C-library level?

**Comment by apolcyn** [comment]
@Gubbi avoiding use grpc library in the parent process before forking is the best thing to do AFAIK. Note that since the change in https://github.com/grpc/grpc/pull/10670, the library won't initialize until the first grpc object (e.g. channel/stub, server), is created.

**Comment by ebenoist** [comment]
Having an explicit reset or shutdown and start hook would make this issue much simpler to deal with. Is it not possible to stop the underlying event loop and call the `grpc_init()` function again? Would exposing `grpc_rb_shutdown` and `grpc_ruby_once_init_internal` allow the underlying library to be reset?

**Comment by apolcyn** [comment]
@ebenoist  something along the lines of what you described should be possible now, and it's actually something I've been meaning to do. What I'm thinking is we can expose global "before fork" and "after fork" hooks which applications are responsible for calling before and after forking, and which will themselves basically shut down and restart the grpc library. We can probably get such an API available within the next couple of releases (not the soon-to-come 1.11 release, but probably in the 1.12 or 1.13 releases).

**Comment by ebenoist** [comment]
@apolcyn Thank you so much for your prompt response. That API makes a lot of sense to me and I'd be eager to test it out for you folks. Please let me know if there is anything I can do to help.

**Comment by blowmage** [comment]
This issue is still outstanding, AFAIK. We still want to be able to reset the Ruby GRPC client's low-level resource after a fork.

**Comment by jeremywadsack** [comment]
This affects us as well using Google Cloud Monitoring (Stackdriver). I am trying to write metics from background Resque jobs. Resque forks from the main process for each job and the jobs all fail with this:

> grpc cannot be used before and after forking

Using `grpc` 1.25.0.

I've looked through the discussions here and I'm not sure that the "fix" in #16332 (https://github.com/grpc/grpc/pull/16332/files#diff-40f6e37e5d9670d49001d5551bc9da82R275) is correct. It checks if the PID has changed which happens when the API forks.

If GRPC is now lazy loaded (I think I saw that), then if we load it on each fork, then shouldn't that work without hanging?

**Comment by blowmage** [comment]
This issue is still outstanding.

**Comment by anujbiyani** [comment]
Copying @blowmage 's bump [from earlier](https://github.com/grpc/grpc/issues/8798#issuecomment-528396349):

> This issue is still outstanding, AFAIK. We still want to be able to reset the Ruby GRPC client's low-level resource after a fork.

**Comment by blowmage** [comment]
This issue is still outstanding, AFAIK. We still want to be able to reset the Ruby GRPC client's low-level resource after a fork.

**Comment by fabirydel** [comment]
Is there any update regarding this issue? I'm using Google Cloud Logging and must absolutely send logs from both the parent and child processes. As it stands, I can't use the `stackdriver` gem because of this issue, has anyone come up with a workaround?

**Comment by jrun** [comment]
@fabirydel If you're systems have google-fluentd installed, you can configure a [forward source](https://docs.fluentd.org/input/forward) that listens on `127.0.0.1`. The fluent [out_google_cloud](https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/blob/master/lib/fluent/plugin/out_google_cloud.rb) has the details on what the JSON payload should be.

**Comment by ianks** [comment]
We are running into this as well. Really painful.

**Comment by matthewford** [comment]
bump

**Comment by syed-mohsin** [comment]
bump :)

**Comment by apolcyn** [comment]
FYI there is work in progress on this in https://github.com/grpc/grpc/pull/33430 (should get into the 1.57 release)

**Comment by apolcyn** [comment]
Reopening because https://github.com/grpc/grpc/pull/33430 isn't a complete fix (forking with bidi streams is not yet supported with that, in particular)

**Comment by ytaben** [comment]
Greetings,

Happy to open a new issue but thought it's relevant to this one (and in case others here want to use puma + refork)

We are working on adding GRPC to our Ruby stack safely and while the pre\post fork hooks work well for the typical pre-fork server model - we have a new challenge with Puma's fork worker feature (reference [here](https://github.com/puma/puma/blob/master/docs/fork_worker.md)).
TL;DR of the feature is that periodically, the worker processes are stopped and re-forked from the _first_ worker process (_not_ the 'cold' master process).
The first process will stop servicing all requests and fork itself so that the application is as hot as possible.

The caveat is that `puma` uses a background thread to listen to a signal that makes the worker stop and fork grand children so we will always fail [this condition](https://github.com/grpc/grpc/blob/1178b2d229ad7fd27df743176dbc2ea104d6b28e/src/ruby/ext/grpc/rb_grpc.c#L380) unless we do some sorcery to signal back to the main thread to _actually_ call fork.

We are wondering about the rationale behind the same-thread requirement given that the server is guaranteed to stop servicing requests and we are seemingly cleaning up _global_ (not thread specific) objects in the prefork hook.
We are also wondering whether it can be removed either entirely or conditionally if the caller can provide certain guarantees like no GRPC access.

CC @apolcyn since you authored the code

Happy to provide additional info and thank you in advance!

**Comment by chjasonwu** [comment]
Following up on the [previous discussion](https://github.com/grpc/grpc/issues/8798#issuecomment-2429591619), I've observed that the same thread is validated only during the prefork and postfork signaling, but not when the channel is initialized or within `grpc_ruby_fork_guard`. I would like to understand why validation happens for one scenario but not the other. Thank you in advance for your assistance. cc @apolcyn
